### PR TITLE
Update snapshot message format

### PR DESCRIPTION
**Pull request by Remagpie**

### CODE REVIEW

File: spec/Block-Synchronization-Extension.md [hunk 1]
```diff
@@ -53,32 +53,17 @@ Request corresponding bodies for each hash.
...
-
-* Identifier: 0x06
-* Restriction: Block number of requested block MUST be multiple of 214.
-
-
 ### GetStateChunk
 
 ```
-GetStateChunk(block_hash, tree_root)
+GetStateChunk(block_hash, chunk_root)
```
Comment (foriequal0) [inline spec/Block-Synchronization-Extension.md]: See the proposal closely. It is a list of chunk roots.
https://research.codechain.io/t/snapshot-sync-proposal/21

File: spec/Block-Synchronization-Extension.md [hunk 2]
```diff
@@ -113,30 +98,15 @@ Response to `GetBodies` message. Snappy algorithm is used to compress content.
...
-
-
 ### StateChunk
 ```
-StateChunk(compressed((key_0, value_0), …) | [])
+StateChunk(compressed([(key_0, value_0), …] | []))
 ```
 
-Response to `GetStateChunk` message. Details of message is same as `StateHead` message.
+Response to `GetStateChunk` message. Key and value included in this messages are raw value stored in state trie. Snappy algorithm is used for compression of content.
```
Comment (foriequal0) [inline spec/Block-Synchronization-Extension.md]:
```suggestion
Response to `GetStateChunk` message. Keys and values included in this messages are raw value stored in state trie. Snappy algorithm is used for compression of content.
```

File: spec/Block-Synchronization-Extension.md [hunk 3]
```diff
@@ -113,30 +98,15 @@ Response to `GetBodies` message. Snappy algorithm is used to compress content.
...
-* Restriction:
-  * State root of requested block MUST be included
-  * For all nodes with depth of less than 2 included in this message, all of its child MUST also be included.
-  * Content MUST be empty array if sender didn’t have requested data
-
-
 ### StateChunk
 ```
-StateChunk(compressed((key_0, value_0), …) | [])
+StateChunk(compressed([(key_0, value_0), …] | []))
```
Comment (foriequal0) [inline spec/Block-Synchronization-Extension.md]: The format is different from the proposal.

File: spec/Block-Synchronization-Extension.md (hunk 1 above)
Comment (Remagpie) [reply spec/Block-Synchronization-Extension.md]: Sorry for the mistake.
I think this format is better though. I left a suggestion to the proposal.

File: sync/src/block/message/response.rs [hunk 4]
```diff
@@ -18,14 +18,14 @@ use rlp::{DecoderError, Encodable, RlpStream, UntrustedRlp};
...
+use cmerkle::snapshot::chunk::RawChunk;
 use ctypes::Header;
 
 #[derive(Debug, PartialEq)]
 pub enum ResponseMessage {
     Headers(Vec<Header>),
     Bodies(Vec<Vec<UnverifiedTransaction>>),
-    StateHead(Vec<u8>),
-    StateChunk(Vec<u8>),
+    StateChunk(Vec<Option<RawChunk>>),
```
Comment (foriequal0) [inline sync/src/block/message/response.rs]: It should be `Vec<u8>`

File: util/merkle/src/snapshot/chunk.rs [hunk 5]
```diff
@@ -59,7 +59,7 @@ impl<'a> std::fmt::Debug for NodeDebugAdaptor<'a> {
 }
 
 /// An unverified chunk from the network
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
```
Comment (foriequal0) [inline util/merkle/src/snapshot/chunk.rs]: Is it necessary?

**Pull request merged**